### DISCUSSION TITLE
Load pages in Storybook

### DISCUSSION THREAD

**Discussion by eric-burel** [vercel/next.js#28290] [Ideas]
### Describe the feature you'd like to request

Being able to create stories for full pages in Storybook.

### Describe the solution you'd like

We need to reuse the Next.js logic that remove server-side code from the client-bundle, otherwise Storybook will also try to bundle packages used in the `getServerSideProps`, `getInitialProps`, `getStaticPaths` and `getStaticProps` functions. The build will thus fail most of the time for pages with data fetching requirements.

The solution would take the form of a reusable Babel/webpack plugin, based on Next.js code for tree shaking that removes the server-side code from the client bundle. This is why I open this ticket in Next.js instead of Storybook, because the solution would be from Next.js side.

### Describe alternatives you've considered

Palliative:
- Create components pages into `src/components/pages`, and load them in the actual pages in `src/pages`. This way, the visual part and the data fetching part of the pages lives in different places. In Storybook, you import only the page component, not the actual page.

Limit is that you often want to simultaneously write data fetching methods and the page component.

`next/head` seems to work out of the box. For `next/router` we may use [Next Router addon](https://storybook.js.org/addons/storybook-addon-next-router), it allows to setup query parameters correctly for instance.

I've also used the recent [storybook-addon-next-router](https://storybook.js.org/addons/storybook-addon-next-router) with great success.

Other approach:
- checkout [Next Page Tester](https://github.com/toomuchdesign/next-page-tester), for Jest. That's the opposite approach: this will actually run the data fetching methods. It could be adapted for Storybook.

**Comment by scottrippey** [comment]
I created `next-router-mock` to handle this, and it works well with Storybook.  You can enable it globally via a decorator, or per-component via a context provider:
https://www.npmjs.com/package/next-router-mock
